src/impl/manifest/objects/nevra/nevraparser.cpp

```diff
@@ -13,40 +14,67 @@ NevraParseError::NevraParseError(const std::string & message)
 NevraParser::NevraParser(std::shared_ptr<INevraFactory> nevra_factory)
     : nevra_factory(std::move(nevra_factory)) {}
 
-// TODO(jkolarik): Refactor to share the common evr logic
+static void splitEVR(std::unique_ptr<INevra> & nevra, const std::string & evr_string) {
+    rpmver evr = NULL;
+    const char *s = NULL;
+    std::string val;
+
+    // split up the epoch:version-release substring the librpm way
+    evr = rpmverParse(evr_string.c_str());
+
+    if (evr == NULL) {
+        throw NevraParseError("Failed to parse epoch:version-release substring: " + evr_string);
+    }
+
+    // collect the components
+    s = rpmverE(evr);
+
+    if (s != NULL) {
+        val = s;
+        nevra->set_epoch(val);
+    }
+
+    s = rpmverV(evr);
+
+    if (s != NULL) {
+        val = s;
+        nevra->set_version(val);
+    }
+
+    s = rpmverR(evr);
+
+    if (s != NULL) {
+        val = s;
+        nevra->set_release(val);
+    }
+
+    // cleanup
+    rpmverFree(evr);
+
+    return;
+}
 
 std::unique_ptr<INevra> NevraParser::parse(const IYamlNode & node) const {
     auto nevra = nevra_factory->create();
     auto nevra_string = node.as_string();
 
-    // Find the architecture
+    // Positions within the substring
     size_t last_dot_position = nevra_string.find_last_of('.');
```

Comment:
If the input string is malformed "foo.bar-version-release" (i.e. it's missing an architecture; dot character is allowed anywere), you will wrongly parse the architecture.

I recommend first locating beginning of EVR as the second hyphen from the end. Then parsing EVR with your new splitEVR(). Then splitting get_release() value into release and architecture by a last dot search in that get_release() value.

Reply:
Working on this.

Reply:
I know, right?